components/menu/menu.js

```diff
@@ -0,0 +1,354 @@
+			return false;
+		}
+		if (window.getComputedStyle(item, null).getPropertyValue('display') === 'none') {
+			return false;
+		}
+		return true;
+	}
+
+	_labelChanged() {
+		this.setAttribute('aria-label', this.label);
+		const returnItem = this._getMenuItemReturn();
+		if (returnItem) {
+			returnItem.setAttribute('text', this.label);
+		}
+	}
+
+	_onKeyDown(e) {
+		const rootTarget = e.composedPath()[0];
+		if (this._items.indexOf(rootTarget) === -1) {
+			return;
+		}
+
+		if (e.keyCode === this._keyCodes.DOWN || e.keyCode === this._keyCodes.UP) {
+			// prevent scrolling when up/down arrows pressed
+			e.preventDefault();
+			e.stopPropagation();
+			if (e.keyCode === this._keyCodes.DOWN) {
+				this._focusNext(rootTarget);
+			} else if (e.keyCode === this._keyCodes.UP) {
+				this._focusPrevious(rootTarget);
+			}
+			return;
+		}
+
+		if (this.childView && e.keyCode === this._keyCodes.LEFT) {
+			e.stopPropagation();
+			this.hide();
+			return;
+		}
+
+	}
+
+	_onKeyPress(e) {
+		if (this._items.indexOf(e.composedPath()[0]) === -1) {
+			return;
+		}
+
+		if (e.keyCode === this._keyCodes.DOWN || e.keyCode === this._keyCodes.UP
+			|| e.keyCode === this._keyCodes.SPACE || e.keyCode === this._keyCodes.ENTER
+			|| e.keyCode === this._keyCodes.ESCAPE) {
+			return;
+		}
+
+		e.stopPropagation();
+
+		const startsWith = function(item, value) {
+			if (item.text && item.text.length > 0 && item.text.substr(0, 1) === value) {
+				return true;
+			}
+			return false;
+		};
+
+		const focusableItems = this._items.filter(this._isFocusable, this);
+		if (!focusableItems || focusableItems.length === 0) {
+			return;
+		}
+
+		const targetItemIndex = focusableItems.indexOf(e.composedPath()[0]);
+
+		const getNextOrFirstIndex = function(itemIndex) {
+			if (itemIndex === focusableItems.length - 1) {
+				return 0;
+			}
+			return itemIndex + 1;
+		}.bind(this);
+
+		/* "charCode" is used instead of "key" due to Safari not supporting */
+		const searchChar = String.fromCharCode(e.charCode);
+
+		let itemIndex = getNextOrFirstIndex(targetItemIndex);
+		while (itemIndex !== targetItemIndex) {
+			const item = focusableItems[itemIndex];
+			if (startsWith(item, searchChar)) {
+				item.focus();
+				return;
+			}
+			itemIndex = getNextOrFirstIndex(itemIndex);
+		}
+
+	}
+
+	_getFirstVisibleItem() {
+		for (let x = 0; x < this._items.length; x++) {
+			if (!this._items[x].hidden) {
+				return this._items[x];
+			}
+		}
+		return null;
+	}
+
+	_getLastVisibleItem() {
+		for (let x = this._items.length - 1; x >= 0; x--) {
+			if (!this._items[x].hidden) {
+				return this._items[x];
+			}
+		}
+		return null;
+	}
+
+	_onMenuItemsChanged() {
+		this._items = this._getMenuItems();
+		if (!this._items || this._items.length === 0) {
+			return;
+		}
+
+		const visibleItems = [];
+
+		for (let i = 0; i < this._items.length; i++) {
+			const item = this._items[i];
+			item.removeAttribute('first');
```

Comment:
Are these `first` and `last` attributes used anymore?  Looks like they are applied below on lines 286 & 287.

Reply:
Yup, they are used in `menu-item-styles`, e.g., https://github.com/BrightspaceUI/core/blob/master/components/menu/menu-item-styles.js#L21